livekit/livekit_egress.go

```diff
@@ -0,0 +1,85 @@
+package livekit
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/livekit/psrpc"
+)
+
+const (
+	MsgBackupUsed               = "At least one file went to backup storage"
+	MsgLimitReached             = "Session limit reached"
+	MsgStartNotReceived         = "Start signal not received"
+	MsgLimitReachedWithoutStart = "Session limit reached before start signal"
+	MsgStoppedBeforeStarted     = "Stop called before pipeline could start"
+
+	EndReasonAPI            = "StopEgress API"
+	EndReasonKilled         = "Process killed"
+	EndReasonSrcClosed      = "Source closed"
+	EndReasonLimitReached   = "Session limit reached"
+	EndReasonStreamsStopped = "All streams stopped"
+	EndReasonFailure        = "Failure"
+)
+
+func (e *EgressInfo) UpdateStatus(status EgressStatus) {
+	e.Status = status
+	e.UpdatedAt = time.Now().UnixNano()
+}
+
+func (e *EgressInfo) SetBackupUsed() {
+	e.Status = EgressStatus_EGRESS_FAILED
+	e.UpdatedAt = time.Now().UnixNano()
+	e.Error = MsgBackupUsed
+	e.ErrorCode = int32(http.StatusPermanentRedirect)
```

Comment:
How will we communicate the case where a segmented file egress eventually ended as failure, but where some segments have failed to upload as well? Customers may want to download whatever segments are available, but will need a way to see that some are only available on the backup storage.

What I'm getting at is that I'm wondering if backup status should be orthogonal to the egress status/error state.

Reply:
True, I think we'll need to add something else because even a separate backup_used state wouldn't cover that